resource/systemd/unit/unit.go

```diff
@@ -0,0 +1,501 @@
+// Copyright © 2016 Asteris, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/asteris-llc/converge/resource"
+	"github.com/asteris-llc/converge/resource/systemd"
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/hashicorp/go-multierror"
+)
+
+// Unit represents a systemd units
+// NOTE: This unit does not parallelize with itself
+type Unit struct {
+	Name          string
+	Active        bool
+	UnitFileState systemd.UnitFileState
+	StartMode     systemd.StartMode
+	Timeout       time.Duration
+
+	resource.TaskStatus
+}
+
+var loadError = `LoadError: "[\"org.freedesktop.DBus.Error.FileNotFound\", \"No such file or directory\"]"`
+
+// Check if all the properties for the unit are correct
+// 1. Checks if the unit is currently loading, if so waits Default 5 seconds
+// 2. Checks if the unit is in the active state
+// 3. Check if the unit is in the unit file state
+func (t *Unit) Check(r resource.Renderer) (resource.TaskStatus, error) {
+	systemd.ApplyDaemonReload()
+	/*First thing to do is to check if the Name given is outside
+	the normal search path of systemd. If so it should be linked
+	*/
+	dir, unitName := filepath.Split(t.Name)
+	shouldBeLinked := false
+	if dir != "" && t.UnitFileState.IsLinked() {
+		// The fullpath to a unit was given. Assume that this isn't in the
+		// normal search paths
+		shouldBeLinked = true
+	}
+
+	// Get the connection from the pool
+	conn, err := systemd.GetDbusConnection()
+	if err != nil {
+		t.TaskStatus = &resource.Status{
+			Level:  resource.StatusFatal,
+			Output: []string{err.Error()},
+		}
+		return t, err
+	}
+	defer conn.Return()
+	dbusConn := conn.Connection
+
+	//Create context
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+	)
+	if t.Timeout == 0 {
+		t.Timeout = systemd.DefaultTimeout
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), t.Timeout)
+	defer cancel()
+
+	// Waits for the daemon to finish loading.
+	// After check for unit linking
+	err = systemd.WaitToLoad(ctx, dbusConn, unitName)
+
+	if err != nil {
```

Comment:
It might make sense to break this out into it's own function since execution will end inside of the block.